crawler/db.py

```diff
@@ -259,3 +268,72 @@ def init_warehouse_db_command():
     mysql_conn.close()
 
     logger.debug("Done")
+
+
+def create_dart_sql_server_conn(config: ModuleType, readonly=True) -> pyodbc.Connection:
+    """Create a SQL Server connection to DART with the given config parameters.
+
+    Arguments:
+        config {ModuleType} -- application config specifying database details
+
+    Returns:
+        pyodbc.Connection -- connection object used to interact with the sql server database
+    """
+    dart_db_host = config.DART_DB_HOST  # type: ignore
+    dart_db_port = config.DART_DB_PORT  # type: ignore
+    if readonly:
+        dart_db_username = config.DART_DB_RO_USER  # type: ignore
+        dart_db_password = config.DART_DB_RO_PASSWORD  # type: ignore
+    else:
+        dart_db_username = config.DART_DB_RW_USER  # type: ignore
+        dart_db_password = config.DART_DB_RW_PASSWORD  # type: ignore
+    dart_db_db = config.DART_DB_DBNAME  # type: ignore
+    dart_db_driver = config.DART_DB_DRIVER  # type: ignore
+
+    connection_string = 'DRIVER='+dart_db_driver+';SERVER='+dart_db_host+f';PORT={dart_db_port};DATABASE='+dart_db_db+';UID='+dart_db_username+';PWD=' + dart_db_password
+
+    logger.debug(f"Attempting to connect to {dart_db_host} on port {dart_db_port}")
+
+    sql_server_conn = None
+    try:
+        sql_server_conn = pyodbc.connect(connection_string)
+
+        if sql_server_conn is not None:
+            logger.debug('DART Connection Successful')
+        else:
+            logger.error('DART Connection Failed')
+
+    except Error as e:
+        logger.error(f"Exception on connecting to DART database: {e}")
+
+    return sql_server_conn
```

Comment:
The return type at the top of this function implies we can expect a pyodbc.Connection however it looks like we can also return None. If:
- The code throws an execption
- pyodbc.connect(connection_string) returns None

This means the should be specifying Optional[pyodbc.Connection]

Also, as best I can tell, on failed connections, pyodbc.connect(connection_string) throws, and should never return None. So the check in the middle can be removed. (Unless I'm missing something, the documentation wasn't great)

I'd probably actually throw an exception in these circumstances instead, but that would also requiring restructuring of the code elsewhere, so probably not something to do now.

Reply:
I think this method was implemented as such to mirror the mysql connection creation method. I agree both that it could be changed so it does not catch the error and return None, instead propagating it to the calling function; and that it probably isn't worth doing this as the functionality works either way, and would mean more code changes. I'm going to leave this as is for now